core/src/main/scala/caliban/schema/ObjectFieldResolver.scala

```diff
@@ -0,0 +1,36 @@
+package caliban.schema
+
+import caliban.execution.Field
+import caliban.schema.Step.{ MetadataFunctionStep, ObjectStep }
+
+import scala.collection.immutable.HashMap
+
+final private class ObjectFieldResolver[R, A](
+  objectName: String,
+  fields: Iterable[(String, A => Step[R])]
+) {
+
+  private val fieldsMap: java.util.HashMap[String, A => Step[R]] = {
+    val map = new java.util.HashMap[String, A => Step[R]]()
+    fields.foreach { case (name, resolve) => map.put(name, resolve) }
+    map
+  }
+
+  def resolve(value: A): Step[R] = MetadataFunctionStep(resolveForField(value, _))
+
+  private def resolveForField(
+    value: A,
+    field: Field
+  ): Step[R] = {
+    val fieldsBuilder = HashMap.newBuilder[String, Step[R]]
```

Comment:
Precompute size based on the distinct fields?

Reply:
Map builders don't expose a constructor to pre-compute their size. I took a look at using `fieldsBuilder.sizeHint(...)` but AFAICT the HashMap builder don't seem to override it (i.e., it's ignored).

Also it seems to me that immutable Maps in Scala don't use initial capacity or load factors. My guess is that it's because they don't need to mutate so their new capacity can be calculated on creation (`.result()`) or when updating them with new elements